compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -2307,6 +2300,37 @@ object Types {
     def tp2: Type
     def isAnd: Boolean
     def derivedAndOrType(tp1: Type, tp2: Type)(implicit ctx: Context): Type  // needed?
+
+    private[this] var myBaseClassesPeriod: Period = Nowhere
+    private[this] var myBaseClasses: List[ClassSymbol] = _
+
+    /** Base classes of And are the merge of the operand base classes
+     *  For OrTypes, it's the intersection.
+     */
+    override final def baseClasses(implicit ctx: Context) = {
+      if (myBaseClassesPeriod != ctx.period) {
+        val bcs1 = tp1.baseClasses
+        val bcs1set = BaseClassSet(bcs1)
+        def recur(bcs2: List[ClassSymbol]): List[ClassSymbol] = bcs2 match {
+          case bc2 :: bcs2rest =>
+            if (isAnd)
+              if (bcs1set contains bc2)
+                if (bc2.is(Trait)) recur(bcs2rest)
+                else bcs1 // common class, therefore rest is the same in both sequences
+              else bc2 :: recur(bcs2rest)
+            else
+              if (bcs1set contains bc2)
+                if (bc2.is(Trait)) bc2 :: recur(bcs2rest)
+                else bcs2rest 
```

Comment:
Shouldn't this be `bcs2` ? Otherwise we miss `bc2` even though it's contained in the baseclasses of both operands